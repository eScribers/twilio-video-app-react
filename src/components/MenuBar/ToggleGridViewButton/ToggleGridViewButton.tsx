import React from 'react';
import GridOnIcon from '@material-ui/icons/GridOn';
import IconButton from '@material-ui/core/IconButton';
import { ViewModeArray, Settings } from '../../../state/settings/settingsReducer';
import { useAppState } from '../../../state';

export default function ToggleGridViewButton() {
  const { settings, dispatchSetting } = useAppState();
  const handleChange = () => {
    const indexOfCurrMode = ViewModeArray.indexOf(settings.viewMode);
    const indexOfNextMode = indexOfCurrMode + 1 < ViewModeArray.length ? indexOfCurrMode + 1 : 0;
    const nextView = ViewModeArray[indexOfNextMode];
    dispatchSetting({
      name: 'viewMode' as keyof Settings,
      value: nextView,
    });
  };

  return (
    <IconButton onClick={handleChange}>
      <GridOnIcon />
    </IconButton>
  );
}
