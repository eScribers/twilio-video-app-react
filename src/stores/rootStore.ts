import rootStore from './makeStore';

export default rootStore;
