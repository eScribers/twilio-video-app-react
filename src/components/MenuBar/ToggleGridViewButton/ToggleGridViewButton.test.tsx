import React from 'react';
import rootStore, { RootStore } from '../../../stores/makeStore';
import ToggleGridViewButton from './ToggleGridViewButton';
import { shallow, mount } from 'enzyme';
// import { initialSettings } from '../../../state/settings/settingsReducer';
import IconButton from '@material-ui/core/IconButton';
import { act } from 'react-dom/test-utils';

jest.mock('../../../stores', () => {
  return {
    __esModule: true, // this property makes it work
    default: rootStore,
  };
});
// import { useAppState } from '../../../hooks/useAppState/useAppState';

// jest.mock('../../../hooks/useAppState/useAppState');
// const mockUseAppState = useAppState as jest.Mock<any>;
// const mockDispatchSetting = jest.fn();
// mockUseAppState.mockImplementation(() => ({ settings: initialSettings, dispatchSetting: mockDispatchSetting }));
describe('the ToggleGridViewButton', () => {
  beforeEach(() => {
    let newStore = new RootStore();
    rootStore.participantsStore = newStore.participantsStore;
    rootStore.roomsStore = newStore.roomsStore;
    global.innerWidth = 1280;
    global.innerHeight = 1024;
    global.dispatchEvent(new Event('resize'));
  });

  it('should render correctly', () => {
    const wrapper = shallow(<ToggleGridViewButton />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render collaboration view', async () => {
    jest.spyOn(rootStore.roomsStore, 'setSetting');
    const wrapper = mount(<ToggleGridViewButton />);
    act(() => {
      // Change the viewport to 750px.
      global.innerWidth = 750;
      // // Trigger the window resize event.
      global.dispatchEvent(new Event('resize'));
    });
    expect(rootStore.roomsStore.setSetting).toHaveBeenCalledWith('viewMode', 'collaboration');
  });

  it('should not change viewMode', async () => {
    jest.spyOn(rootStore.roomsStore, 'setSetting');
    const wrapper = shallow(<ToggleGridViewButton />);
    act(() => {
      // Change the viewport to 750px.
      global.innerWidth = 800;
      // // Trigger the window resize event.
      global.dispatchEvent(new Event('resize'));
    });
    expect(rootStore.roomsStore.setSetting).not.toHaveBeenCalled();
  });

  it('should render the next view mode', () => {
    jest.spyOn(rootStore.roomsStore, 'setSetting');
    const wrapper = shallow(<ToggleGridViewButton />);
    act(() => {
      wrapper.find(IconButton).simulate('click');
    });
    expect(rootStore.roomsStore.setSetting).toHaveBeenCalledWith('viewMode', 'grid 2 column');
  });
});
