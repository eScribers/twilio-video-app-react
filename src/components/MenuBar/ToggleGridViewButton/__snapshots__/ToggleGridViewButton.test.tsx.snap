// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ToggleGridViewButton should render correctly 1`] = `
<WithStyles(ForwardRef(IconButton))
  onClick={[Function]}
>
  <Memo([object Object]) />
</WithStyles(ForwardRef(IconButton))>
`;
